jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocInfo.java

```diff
@@ -226,6 +242,24 @@ public Cluster findClusterInPool(String filter) throws SQLException {
         }
     }
 
+    /**
+     * Random weighted selection of a cluster based on clusters' available yarn memory.
+     *
+     * @param clusterLoads map from cluster to the available yarn memory of that cluster
+     * @return the selected cluster
+     */
+    public static Cluster pickCluster(Stream<Map.Entry<Cluster, Long>> clusterLoads) {
+        Random random = new Random();
+        return clusterLoads
+                .map(
+                        e ->
+                                new AbstractMap.SimpleEntry<>(
+                                        e.getKey(), -Math.log(random.nextDouble()) / e.getValue()))
+                .min(Comparator.comparing(AbstractMap.SimpleEntry::getValue))
+                .orElseThrow(IllegalArgumentException::new)
```

Comment:
Does this just happen if clusterLoads is empty? It might be worth asserting that clusterLoads is not empty, and throwing SQLException if no clusters are present